lib/stellar/sep10.rb

```diff
@@ -0,0 +1,353 @@
+    ] => C::ArrayOf[Stellar::AccountSigner])
+    # Verifies that for a SEP 10 challenge transaction all signatures on the transaction 
+    # are accounted for and that the signatures meet a threshold on an account. A 
+    # transaction is verified if it is signed by the server account, and all other 
+    # signatures match a signer that has been provided as an argument, and those 
+    # signatures meet a threshold on the account.
+    #
+    # @param challenge_transaction [String] SEP0010 transaction challenge transaction in base64.
+    # @param server [Stellar::KeyPair] keypair for server's account.
+    # @param threshold [Integer] The medThreshold on the client account.
+    # @param signers [ArrayOf[Stellar::AccountSigner]]The signers of client account.
+    #
+    # @return [ArrayOf[Stellar::AccountSigner]]
+    #
+    # Raises a InvalidSep10ChallengeError if:
+    #   - The transaction is invalid according to Stellar::SEP10.read_challenge_transaction.
+    #   - One or more signatures in the transaction are not identifiable as the server 
+    #     account or one of the signers provided in the arguments.
+    #   - The signatures are all valid but do not meet the threshold.
+    def self.verify_challenge_transaction_threshold(
+      challenge_transaction:,
+      server:,
+      threshold:,
+      signers:
+    )
+      signers_found = verify_challenge_transaction_signers(
+        challenge: challenge_transaction, 
+        server: server, 
+        signers: signers
+      )
+  
+      weight = 0
+      signers_found.each do |s|
+        weight += s.weight
+      end
+
+      if weight < threshold
+        raise InvalidSep10ChallengeError.new(
+          "signers with weight %d do not meet threshold %d." % [weight, threshold]
+        )
+      end
+  
+      return signers_found
+    end
+
+    Contract(C::KeywordArgs[
+      challenge_transaction: String, 
+      server: Stellar::KeyPair
+    ] => nil)
+    # Verifies if a transaction is a valid per SEP-10 challenge transaction, if the validation 
+    # fails, an exception will be thrown.
+    #
+    # This function performs the following checks:
+    #   1. verify that transaction sequenceNumber is equal to zero;
+    #   2. verify that transaction source account is equal to the server's signing key;
+    #   3. verify that transaction has time bounds set, and that current time is between the minimum and maximum bounds;
+    #   4. verify that transaction contains a single Manage Data operation and it's source account is not null;
+    #   5. verify that transaction envelope has a correct signature by server's signing key;
+    #   6. verify that transaction envelope has a correct signature by the operation's source account;
+    #
+    # @param challenge_transaction [String] SEP0010 transaction challenge transaction in base64.
+    # @param server [Stellar::KeyPair] keypair for server's account.
+    #
+    # Raises a InvalidSep10ChallengeError if the validation fails
```

Comment:
Can we add a deprecated notice here and point people to using the threshold and signer functions?

Reply:
Why don't I just remove it if it should be deprecated as soon as its introduced?

Reply:
To avoid making this a breaking change, plus implementers may choose to support the old and new versions of SEP-10 for a time and leaving the existing functions may help that.